/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

syntax = "proto3";
option java_generic_services = true;
option java_package = "com.aiflow.notification.proto";
option go_package = "/notification_service";
option py_generic_services = true;

import "google/api/annotations.proto";

package notification_service;

service NotificationService {
    // AirFlowService provides notification function rest endpoint of NotificationService for Notification Service component.
    // Functions of NotificationService include:
    //  1.Send event.
    //  2.List events.

    // Send event.
    rpc sendEvent (SendEventRequest) returns (SendEventsResponse) {
        option (google.api.http) = {
            post: "/aiflow/notification-service/send_event"
            body: "*"
        };
    }

    // List events.
    rpc listEvents (ListEventsRequest) returns (ListEventsResponse) {
        option (google.api.http) = {
            post: "/aiflow/notification-service/list_events"
            body: "*"
        };
    }

    // List all events
    rpc listAllEvents (ListAllEventsRequest) returns (ListEventsResponse) {
        option (google.api.http) = {
            post: "/aiflow/notification-service/list_all_events"
            body: "*"
        };
    }

    // Accepts notifications from other members.
    rpc notify (NotifyRequest) returns (NotifyResponse) {
        option (google.api.http) = {
            post: "/aiflow/notification-service/notify"
            body: "*"
        };
    }

    // List current living members.
    rpc listMembers (ListMembersRequest) returns (ListMembersResponse) {
        option (google.api.http) = {
            post: "/aiflow/notification-service/list_members"
            body: "*"
        };
    }

    // Notify current members that there is a new member added.
    rpc notifyNewMember (NotifyNewMemberRequest) returns (NotifyNewMemberResponse) {
        option (google.api.http) = {
            post: "/aiflow/notification-service/notify_new_member"
            body: "*"
        };
    }

    // Get latest version by key
    rpc getLatestVersionByKey (GetLatestVersionByKeyRequest) returns (GetLatestVersionResponse) {
        option (google.api.http) = {
            post: "/aiflow/notification-service/get_latest_version_by_key"
            body: "*"
        };
    }
}
message EventProto {
    string key = 1;
    string value = 2;
    string event_type = 3;
    string context = 4;
    string namespace = 5;
    int64 version = 6;
    int64 create_time = 7;
    string sender = 8;
}


message SendEventRequest {
    EventProto event = 1;
    // use uuid to identify retry
    string uuid = 2;
}

enum ReturnStatus {
    SUCCESS = 0;
    ERROR = 1;
}

message SendEventsResponse {
    EventProto event = 1;
    ReturnStatus return_code = 2;
    string return_msg = 3;
}

message ListEventsRequest {
    repeated string keys = 1;
    string event_type = 2;
    int64 start_time = 3;
    int64 start_version = 4;
    int32 timeout_seconds = 5;
    string namespace = 6;
    string sender = 7;
}

message ListAllEventsRequest {
    int32 timeout_seconds = 1;
    int64 start_time = 2;
    int64 start_version = 3;
    int64 end_version = 4;
}

message ListEventsResponse {
    ReturnStatus return_code = 1;
    string return_msg = 2;
    repeated EventProto events = 3;
}

message GetLatestVersionByKeyRequest {
    string key = 1;
    string namespace = 2;
    int32 timeout_seconds = 3;
}

message GetLatestVersionResponse {
    string return_code = 1;
    string return_msg = 2;
    int64 version = 3;
}

message MemberProto {
    int64 version = 1;
    string server_uri = 2;
    int64 update_time = 3;
}

message Notify {
    string key = 1;
    string namespace = 2;
}

message NotifyRequest {
    repeated Notify notifies = 1;
}

message NotifyResponse {
    ReturnStatus return_code = 1;
    string return_msg = 2;
}

message ListMembersRequest {
    int32 timeout_seconds = 1;
}

message ListMembersResponse {
    ReturnStatus return_code = 1;
    string return_msg = 2;
    repeated MemberProto members = 3;
}

message NotifyNewMemberRequest {
    MemberProto member = 1;
}

message NotifyNewMemberResponse {
    ReturnStatus return_code = 1;
    string return_msg = 2;
}
